File: android/app/src/main/java/app/droidmatch/m1/SessionAuthenticator.java
package app.droidmatch.m1;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * Cross-platform canonical transcript and HMAC/HKDF core for paired sessions.
 *
 * <p>This class deliberately contains no protobuf, storage, or UI state. Swift and
 * Java must match the same checked-in fixture before the authentication messages are
 * connected to {@link RpcDispatcher}.</p>
 */
public final class SessionAuthenticator {
    public static final int PAIRING_ID_LENGTH = 16;
    public static final int NONCE_LENGTH = 32;
    public static final int PAIRING_KEY_LENGTH = 32;
    public static final int DIGEST_LENGTH = 32;

    private static final byte[] TRANSCRIPT_CONTEXT = ascii("DroidMatch session auth v1\0");
    private static final byte[] CLIENT_PROOF_CONTEXT = ascii("DroidMatch client proof v1\0");
    private static final byte[] SERVER_PROOF_CONTEXT = ascii("DroidMatch server proof v1\0");
    private static final byte[] SESSION_KEY_CONTEXT = ascii("DroidMatch session key v1\0");

    private SessionAuthenticator() {
    }

    public static byte[] transcript(
            byte[] pairingId,
            byte[] clientNonce,
            byte[] serverNonce,
            int protocolMajor,
            int protocolMinor,
            int transportKind
    ) {
        requireLength(pairingId, "pairing ID", PAIRING_ID_LENGTH);
        requireLength(clientNonce, "client nonce", NONCE_LENGTH);
        requireLength(serverNonce, "server nonce", NONCE_LENGTH);

        ByteArrayOutputStream output = new ByteArrayOutputStream();
        append(output, TRANSCRIPT_CONTEXT);
        appendLengthPrefixed(output, pairingId);
        appendLengthPrefixed(output, clientNonce);
        appendLengthPrefixed(output, serverNonce);
        appendUInt32(output, protocolMajor);
        appendUInt32(output, protocolMinor);
        appendUInt32(output, transportKind);
        return output.toByteArray();
    }

    public static byte[] transcriptHash(byte[] transcript) {
        try {
            return MessageDigest.getInstance("SHA-256").digest(transcript);
        } catch (GeneralSecurityException exception) {
            throw new IllegalStateException("SHA-256 is unavailable", exception);
        }
    }

    public static byte[] clientProof(byte[] pairingKey, byte[] transcriptHash) {
        return proof(pairingKey, transcriptHash, CLIENT_PROOF_CONTEXT);
    }

    public static byte[] serverProof(byte[] pairingKey, byte[] transcriptHash) {
        return proof(pairingKey, transcriptHash, SERVER_PROOF_CONTEXT);
    }

    public static boolean verifyClientProof(
            byte[] candidate,
            byte[] pairingKey,
            byte[] transcriptHash
    ) {
        return verifyProof(candidate, pairingKey, transcriptHash, CLIENT_PROOF_CONTEXT);
    }

    public static boolean verifyServerProof(
            byte[] candidate,
            byte[] pairingKey,
            byte[] transcriptHash
    ) {
        return verifyProof(candidate, pairingKey, transcriptHash, SERVER_PROOF_CONTEXT);
    }

    public static byte[] sessionKey(byte[] pairingKey, byte[] transcriptHash) {
        requireLength(pairingKey, "pairing key", PAIRING_KEY_LENGTH);
        requireLength(transcriptHash, "transcript hash", DIGEST_LENGTH);
        byte[] pseudoRandomKey = hmac(transcriptHash, pairingKey);
        byte[] expandInput = new byte[SESSION_KEY_CONTEXT.length + 1];
        System.arraycopy(SESSION_KEY_CONTEXT, 0, expandInput, 0, SESSION_KEY_CONTEXT.length);
        expandInput[expandInput.length - 1] = 1;
        try {
            return hmac(pseudoRandomKey, expandInput);
        } finally {
            java.util.Arrays.fill(pseudoRandomKey, (byte) 0);
            java.util.Arrays.fill(expandInput, (byte) 0);
        }
    }

    private static byte[] proof(byte[] pairingKey, byte[] transcriptHash, byte[] roleContext) {
        requireLength(pairingKey, "pairing key", PAIRING_KEY_LENGTH);
        requireLength(transcriptHash, "transcript hash", DIGEST_LENGTH);
        return hmac(pairingKey, concatenate(roleContext, transcriptHash));
    }

    private static boolean verifyProof(
            byte[] candidate,
            byte[] pairingKey,
            byte[] transcriptHash,
            byte[] roleContext
    ) {
        requireLength(pairingKey, "pairing key", PAIRING_KEY_LENGTH);
        requireLength(transcriptHash, "transcript hash", DIGEST_LENGTH);
        if (candidate.length != DIGEST_LENGTH) {
            return false;
        }
        return MessageDigest.isEqual(proof(pairingKey, transcriptHash, roleContext), candidate);
    }

    private static byte[] hmac(byte[] key, byte[] message) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key, "HmacSHA256"));
            return mac.doFinal(message);
        } catch (GeneralSecurityException exception) {
            throw new IllegalStateException("HmacSHA256 is unavailable", exception);
        }
    }

    private static byte[] concatenate(byte[] first, byte[] second) {
        byte[] result = new byte[first.length + second.length];
        System.arraycopy(first, 0, result, 0, first.length);
        System.arraycopy(second, 0, result, first.length, second.length);
        return result;
    }

    private static void requireLength(byte[] value, String field, int expected) {
        if (value.length != expected) {
            throw new IllegalArgumentException(
                    "invalid " + field + " length: expected " + expected + " bytes, got " + value.length
            );
        }
    }

    private static void appendLengthPrefixed(ByteArrayOutputStream output, byte[] value) {
        int length = value.length;
        output.write((length >>> 8) & 0xff);
        output.write(length & 0xff);
        append(output, value);
    }

    private static void appendUInt32(ByteArrayOutputStream output, int value) {
        output.write((value >>> 24) & 0xff);
        output.write((value >>> 16) & 0xff);
        output.write((value >>> 8) & 0xff);
        output.write(value & 0xff);
    }

    private static void append(ByteArrayOutputStream output, byte[] value) {
        output.write(value, 0, value.length);
    }

    private static byte[] ascii(String value) {
        return value.getBytes(StandardCharsets.US_ASCII);
    }
}

File: mac/Sources/DroidMatchCore/SessionAuthenticator.swift
import CryptoKit
import Foundation

public enum SessionAuthenticationError: Error, CustomStringConvertible, Sendable {
    case invalidLength(field: String, expected: Int, actual: Int)

    public var description: String {
        switch self {
        case let .invalidLength(field, expected, actual):
            return "invalid \(field) length: expected \(expected) bytes, got \(actual)"
        }
    }
}

/// Cross-platform session-authentication primitives.
///
/// This type intentionally owns only canonical bytes and standard SHA-256/HMAC/HKDF
/// operations. Pairing storage, protobuf messages, retries, and UI confirmation stay
/// outside so Swift and Java can verify this security-critical core with one fixture.
public enum SessionAuthenticator {
    public static let pairingIDLength = 16
    public static let nonceLength = 32
    public static let pairingKeyLength = 32
    public static let digestLength = 32

    private static let transcriptContext = Data("DroidMatch session auth v1\0".utf8)
    private static let clientProofContext = Data("DroidMatch client proof v1\0".utf8)
    private static let serverProofContext = Data("DroidMatch server proof v1\0".utf8)
    private static let sessionKeyContext = Data("DroidMatch session key v1\0".utf8)

    public static func transcript(
        pairingID: Data,
        clientNonce: Data,
        serverNonce: Data,
        protocolMajor: UInt32,
        protocolMinor: UInt32,
        transport: Droidmatch_V1_TransportKind
    ) throws -> Data {
        try requireLength(pairingID, field: "pairing ID", expected: pairingIDLength)
        try requireLength(clientNonce, field: "client nonce", expected: nonceLength)
        try requireLength(serverNonce, field: "server nonce", expected: nonceLength)

        var result = transcriptContext
        appendLengthPrefixed(pairingID, to: &result)
        appendLengthPrefixed(clientNonce, to: &result)
        appendLengthPrefixed(serverNonce, to: &result)
        appendUInt32(protocolMajor, to: &result)
        appendUInt32(protocolMinor, to: &result)
        appendUInt32(UInt32(transport.rawValue), to: &result)
        return result
    }

    public static func transcriptHash(_ transcript: Data) -> Data {
        Data(SHA256.hash(data: transcript))
    }

    public static func clientProof(pairingKey: Data, transcriptHash: Data) throws -> Data {
        try proof(
            pairingKey: pairingKey,
            transcriptHash: transcriptHash,
            roleContext: clientProofContext
        )
    }

    public static func serverProof(pairingKey: Data, transcriptHash: Data) throws -> Data {
        try proof(
            pairingKey: pairingKey,
            transcriptHash: transcriptHash,
            roleContext: serverProofContext
        )
    }

    public static func verifyClientProof(
        _ candidate: Data,
        pairingKey: Data,
        transcriptHash: Data
    ) throws -> Bool {
        try verifyProof(
            candidate,
            pairingKey: pairingKey,
            transcriptHash: transcriptHash,
            roleContext: clientProofContext
        )
    }

    public static func verifyServerProof(
        _ candidate: Data,
        pairingKey: Data,
        transcriptHash: Data
    ) throws -> Bool {
        try verifyProof(
            candidate,
            pairingKey: pairingKey,
            transcriptHash: transcriptHash,
            roleContext: serverProofContext
        )
    }

    public static func sessionKey(pairingKey: Data, transcriptHash: Data) throws -> Data {
        try requireLength(pairingKey, field: "pairing key", expected: pairingKeyLength)
        try requireLength(transcriptHash, field: "transcript hash", expected: digestLength)
        let derived = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: pairingKey),
            salt: transcriptHash,
            info: sessionKeyContext,
            outputByteCount: pairingKeyLength
        )
        return derived.withUnsafeBytes { Data($0) }
    }

    private static func proof(
        pairingKey: Data,
        transcriptHash: Data,
        roleContext: Data
    ) throws -> Data {
        try requireLength(pairingKey, field: "pairing key", expected: pairingKeyLength)
        try requireLength(transcriptHash, field: "transcript hash", expected: digestLength)
        var message = roleContext
        message.append(transcriptHash)
        let code = HMAC<SHA256>.authenticationCode(
            for: message,
            using: SymmetricKey(data: pairingKey)
        )
        return Data(code)
    }

    private static func verifyProof(
        _ candidate: Data,
        pairingKey: Data,
        transcriptHash: Data,
        roleContext: Data
    ) throws -> Bool {
        try requireLength(pairingKey, field: "pairing key", expected: pairingKeyLength)
        try requireLength(transcriptHash, field: "transcript hash", expected: digestLength)
        guard candidate.count == digestLength else {
            return false
        }
        var message = roleContext
        message.append(transcriptHash)
        return HMAC<SHA256>.isValidAuthenticationCode(
            candidate,
            authenticating: message,
            using: SymmetricKey(data: pairingKey)
        )
    }

    private static func requireLength(_ data: Data, field: String, expected: Int) throws {
        guard data.count == expected else {
            throw SessionAuthenticationError.invalidLength(
                field: field,
                expected: expected,
                actual: data.count
            )
        }
    }

    private static func appendLengthPrefixed(_ data: Data, to result: inout Data) {
        let length = UInt16(data.count)
        result.append(UInt8((length >> 8) & 0xff))
        result.append(UInt8(length & 0xff))
        result.append(data)
    }

    private static func appendUInt32(_ value: UInt32, to result: inout Data) {
        result.append(UInt8((value >> 24) & 0xff))
        result.append(UInt8((value >> 16) & 0xff))
        result.append(UInt8((value >> 8) & 0xff))
        result.append(UInt8(value & 0xff))
    }
}
